app/Components/SudokuBoard/SudokuBoard.tsx

```diff
@@ -744,6 +1083,185 @@
+
+        setSudokuHint({
+          ...sudokuHint,
+          stage: sudokuHint.stage + stageOffset,
+        });
+      }
+    }
+
+    const currentStage = sudokuHint.stage + stageOffset; // keep track of updated state
+
+    const removals: number[][] = JSON.parse(
+      JSON.stringify(sudokuHint.hint.removals)
+    );
+
+    if (
+      sudokuHint.hint.strategy === "AMEND_NOTES" &&
+      (currentStage === 4 || currentStage === 5)
+    ) {
+      const r = sudokuHint.hint.removals[0][0];
+      const c = sudokuHint.hint.removals[0][1];
+      const removals = [...sudokuHint.hint.removals[0]]; // deep clone to prevent sudokuHint state update
+      removals.splice(0, 2);
+
+      let notesWereUpdated = false;
+
+      const allNotes = [1, 2, 3, 4, 5, 6, 7, 8, 9];
+      let newNotes: number[] = [];
+      if (sudokuBoard.puzzle[r][c].type == "note") {
+        newNotes = [...(sudokuBoard.puzzle[r][c].entry as number[])];
+      }
+      // Insert missing notes due to AMEND_NOTES hint
+      if (currentStage === 4) {
+        for (const note of allNotes) {
+          if (!removals.includes(note) && !newNotes.includes(note)) {
+            newNotes.push(note);
+            notesWereUpdated = true;
+          }
+        }
+      }
+      // Remove unnecessary notes due to AMEND_NOTES hint
+      else if (currentStage === 5 && sudokuBoard.puzzle[r][c].type == "note") {
+        newNotes = sudokuBoard.puzzle[r][c].entry as number[];
+        for (const note of removals) {
+          if (newNotes.includes(note)) {
+            const index = newNotes.indexOf(note);
+            newNotes.splice(index, 1);
+          }
+        }
+      }
+
+      if (notesWereUpdated) {
+        replaceSudokuBoardCells(
+          [{ type: "note", entry: newNotes }],
+          [{ c: c, r: r }]
+        );
+      }
+    } else if (
+      sudokuHint.hint.strategy === "NAKED_SINGLE" &&
+      currentStage === 5
+    ) {
+      const r = sudokuHint.hint.placements[0][0];
+      const c = sudokuHint.hint.placements[0][1];
+      const placements = [...sudokuHint.hint.placements[0]]; // deep clone to prevent sudokuHint state update
+      placements.splice(0, 2);
+
+      replaceSudokuBoardCells(
+        [{ type: "value", entry: placements[0] }],
+        [{ c: c, r: r }]
+      );
+    } else if (currentStage === 5) {
+      const cells: CellProps[] = [];
+      const locations: CellLocation[] = [];
+      for (const removal of removals) {
+        const r = removal[0];
+        const c = removal[1];
+        removal.splice(0, 2);
+
+        let newNotes = [...(sudokuBoard.puzzle[r][c].entry as number[])];
+        for (const note of removal) {
+          if (newNotes.includes(note)) {
+            const index = newNotes.indexOf(note);
+            newNotes.splice(index, 1);
+          }
+        }
+        cells.push({ type: "note", entry: newNotes });
+        locations.push({ c: c, r: r });
+      }
+      replaceSudokuBoardCells(cells, locations);
+    }
+  };
+
```

Comment:
Do you have any suggestions on how to update this function?

Reply:
I tried this implementation, and it did not work. This will hopefully be a temporary function as we are planning on reimplementing the hint api. I don't think it is worth it at this time to improve this function.